### PR TITLE
Update CTCGreedyDecoder MO operation implementation

### PR DESCRIPTION
**Pull request by achetver**
Description: Add CTCGreedyDecoder operation extender in MO IR Reader. Also update shape infer function for CTCGreedyDecoder operation and extender for ShapeOf operation.

JIRA: CVS-31610

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [ ]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [ ]  Unit tests
* [ ]  Framework operation tests
* [ ]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: Not needed - nothing to check;
* [ ]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list: Not needed - operation already supported;
* [x]  Supported **public** models list: Not needed - no new models supported;
* [x]  New operations specification: Not needed - no new operations supported;
* [x]  Guide on how to convert the **public** model: Not needed - no new models supported;
* [x]  User guide update: Not needed - nothing to update.

### CODE REVIEW

File: model-optimizer/extensions/ops/ctc_greedy_decoder_test.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def test_tf_concat_infer(self):
                                 'ctc': {'ctc_merge_repeated': 1}
                             })
 
+        graph.graph['layout'] = 'NHWC'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/ctc_greedy_decoder_test.py:49]: Please, add a new unit test for "NCHW" layout

File: model-optimizer/extensions/ops/ctc_greedy_decoder.py [hunk 2]
```diff
@@ -47,4 +47,7 @@ def ctc_greedy_decoder_infer(node: Node):
         outn.shape = np.ones(4, dtype=np.int)
         assert inn.shape[1] == inn2.shape[1], 'Batch for CTCGreedyDecoder should be the same in both inputs'
         outn.shape[0] = inn.shape[1]
-        outn.shape[1] = inn.shape[0]
+        if node.graph.graph['layout'] == 'NHWC':
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/ctc_greedy_decoder.py:50]: Looks like an ugly solution. Also it is not clear why we solve it here that way. There should be at least some comment describing how it works.

File: model-optimizer/mo/utils/ir_reader/extenders/ctc_greedy_decoder_extender.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from mo.utils.graph import Node
+from mo.utils.ir_reader.extender import Extender
+
+
+class CTCGreedyDecoder_extender(Extender):
+    op = 'CTCGreedyDecoder'
+
+    @staticmethod
+    def extend(op: Node):
+        if op.graph.graph['cmd_params'].framework == 'tf':
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/ir_reader/extenders/ctc_greedy_decoder_extender.py:28]: Extender is always work with OpenVINO representation of the operations. So there should be no any condition like this. The shape infer function should work according to the operation specification.

**Comment by lazarevevgeny** [comment]
invalid template

**Comment by achetver** [comment]
> invalid template

Template updated, PR converted to draft.

File: model-optimizer/extensions/front/tf/CTCGreedyDecoder.py [hunk 4]
```diff
@@ -61,22 +70,49 @@ def replace_sub_graph(self, graph: Graph, match: dict):
...
+
+        unsqueeze_1 = create_op_node_with_second_input(graph, Unsqueeze, int64_array([0, 1]),
+                                                       {'name': decoder_name + '/unsqueeze_1'})
+        unsqueeze_1.in_port(0).connect(ss.out_port(0))
+
+        unsqueeze_2 = create_op_node_with_second_input(graph, Unsqueeze, int64_array([0]),
+                                                       {'name': decoder_name + '/unsqueeze_2'})
+
+        port = decoder_node.in_port(1).get_source()
+        port.disconnect()
```
Comment (jane-intel) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoder.py:98]: shouldn't we use set/get source to reconnect the edges?

File: model-optimizer/extensions/front/tf/CTCGreedyDecoder.py [hunk 5]
```diff
@@ -61,22 +70,49 @@ def replace_sub_graph(self, graph: Graph, match: dict):
...
+                                                {'name': decoder_name + '/broadcast'})
+
+        squeeze = create_op_node_with_second_input(graph, Squeeze, int64_array([0]),
+                                                   {'name': decoder_name + '/squeeze'})
+        concat.out_port(0).connect(squeeze.in_port(0))
+        squeeze.out_port(0).connect(broadcast.in_port(1))
+
+        broadcast.out_port(0).connect(decoder_node.in_port(1))
+
+        return {}
```
Comment (jane-intel) [inline model-optimizer/extensions/front/tf/CTCGreedyDecoder.py:118]: do we really need to use FrontReplacementSubgraph here? FrontReplacementPattern would be enough. 
Lets handle `nodes_to_remove` business inside the `replace_pattern`.

File: model-optimizer/extensions/ops/ctc_greedy_decoder.py [hunk 6]
```diff
@@ -41,10 +42,10 @@ def supported_attrs(self):
...
     def ctc_greedy_decoder_infer(node: Node):
-        outn = node.out_node(0)
-        inn = node.in_node(0)
-        inn2 = node.in_node(1)
-        outn.shape = np.ones(4, dtype=np.int)
-        assert inn.shape[1] == inn2.shape[1], 'Batch for CTCGreedyDecoder should be the same in both inputs'
-        outn.shape[0] = inn.shape[1]
-        outn.shape[1] = inn.shape[0]
+        output_shape = np.ones(4, dtype=np.int)
+        assert node.in_port(0).data.get_shape()[1] == node.in_port(0).data.get_shape()[1], \
```
Comment (jane-intel) [inline model-optimizer/extensions/ops/ctc_greedy_decoder.py:46]:
```suggestion
        assert node.in_port(0).data.get_shape()[1] == node.in_port(1).data.get_shape()[1], \
```

File: model-optimizer/extensions/ops/ctc_greedy_decoder.py [hunk 7]
```diff
@@ -41,10 +42,10 @@ def supported_attrs(self):
 
     @staticmethod
     def ctc_greedy_decoder_infer(node: Node):
```
Comment (jane-intel) [inline model-optimizer/extensions/ops/ctc_greedy_decoder.py:44]: Lets call it `infer` to unify with other ops

File: model-optimizer/extensions/ops/ctc_greedy_decoder_test.py [hunk 8]
```diff
@@ -23,32 +23,40 @@
...
+                    'ctc_data': {'kind': 'data', 'value': None, 'shape': None},
                     'node_3': {'type': 'Identity', 'kind': 'op'},
+                    'node_3_data': {'kind': 'data', 'value': None, 'shape': None},
                     'op_output': { 'kind': 'op', 'op': 'Result'},
                     }
 
 
 class TestConcatPartialInfer(unittest.TestCase):
-    def test_tf_concat_infer(self):
+    def test_tf_ctc_greedy_decoder_infer(self):
```
Comment (jane-intel) [inline model-optimizer/extensions/ops/ctc_greedy_decoder_test.py:38]: Should we delete this test? unit here is too small.

**Pull request closed without merging**